carpooling/models/models.py

```diff
@@ -0,0 +1,111 @@
+    destination_loc = fields.Char(string="Destination location", required=True)
+    departure_time = fields.Datetime(string="Departure time", required=True)
+    expired = fields.Boolean(compute="_compute_expired", store=True)
+    @api.depends("departure_time")
+    def _compute_expired(self):
+        for record in self:
+            now = datetime.now()
+            record.expired = record.departure_time < now
+                
+
+    available_seats = fields.Integer(required=True, string="Available seats")
+    remaining_seats = fields.Char(compute='_remaining_seats', string="Remaining seats")
+    remaining_seats_int = fields.Integer(compute='_remaining_seats_int', string="Remaining seats")
+    
+    description = fields.Text()
+
+    passengers = fields.Many2many(comodel_name="res.users", string="Passengers")
+
+    @api.constrains('passengers')
+    def _check_number_of_passengers(self):
+        for record in self:
+            if len(record.passengers) > record.available_seats:
+                raise ValidationError(f"Too many passengers ({len(record.passengers)} > {record.remaining_seats_int})")
+
+    # button_txt = fields.Char(compute="_compute_btn_txt")
+
+    # @api.depends("passengers")
+    # def _compute_btn_txt(self):
+    #     for record in self:
+    #         if self.env.user in record.passengers:
+    #             record.button_txt = "Cancel booking"
+    #         else:
+    #             record.button_txt = "Book trip"
+
+    current_user_is_passenger = fields.Boolean(compute="_compute_current_user_is_passenger")
+    @api.depends("passengers")
+    def _compute_current_user_is_passenger(self):
+        for record in self:
+            record.current_user_is_passenger = (self.env.user in record.passengers)
+
+    def book_or_cancel(self):
+        for record in self:
+            if self.env.user in record.passengers:
+                record.write({'passengers': [fields.Command.unlink(self.env.uid)] })  
+            else:
+                if record.remaining_seats_int <= 0:
+                    continue
+                record.write({'passengers': [fields.Command.link(self.env.uid)] })  
+        return True
+    
+    # @api.onchange('passengers')
+    # def _on_passengers_change(self):
+    #     return {
+    #         'warning': {
+    #             'title': "You should NOT do that",
+    #             'message': "The passengers list is NOT editable",
+    #         }
+    #     }
+
+    @api.depends("available_seats", "passengers")
+    def _remaining_seats(self):
+        for record in self:
+            record.remaining_seats = str(record.available_seats - len(record.passengers)) + " out of " + str(record.available_seats)
+
+    @api.depends("available_seats", "passengers")
+    def _remaining_seats_int(self):
+        for record in self:
+            record.remaining_seats_int = record.available_seats - len(record.passengers)
+
+class Passenger(models.Model):
```

Comment:
Generally we use one file / class ( sometime we can mix two if linked together )